spec/unit/forge/module_release_spec.rb

```diff
@@ -7,35 +7,110 @@
-      expect(subject.forge_release).to receive(:download).with(download_path)
-      subject.download
+  context "with cached tarball" do
+    describe '#download' do
+      it "does not download a new tarball" do
+        expect(tarball_cache_path).to receive(:exist?).and_return(true)
+        expect(subject.forge_release).not_to receive(:download).with(download_path)
+        subject.download
+      end
     end
   end
 
   describe '#verify' do
-    it "verifies the module checksum based on the Forge file checksum" do
-      allow(subject.forge_release).to receive(:file_md5).and_return('something')
-      expect(subject.forge_release).to receive(:verify).with(download_path)
+
+    it "verifies using the file md5, if that exists" do
+      allow(File).to receive(:read).and_return(file_contents)
+      allow(md5_digest_class).to receive(:hexdigest).and_return(md5_of_tarball)
+      allow(md5_file_path).to receive(:exist?).and_return(true)
+      expect(subject).to receive(:verify_from_md5_file).with(md5_of_tarball)
       subject.verify
     end
+
+    it "verifies using the forge file_md5, if no md5 file exists" do
+      allow(File).to receive(:read).and_return(file_contents)
+      allow(md5_digest_class).to receive(:hexdigest).and_return(md5_of_tarball)
+      allow(md5_file_path).to receive(:exist?).and_return(false)
+      expect(subject).to receive(:verify_from_forge).with(md5_of_tarball)
+      subject.verify
+    end
+  end
+
+  describe '#verify_from_md5_file' do
+
+    it "does nothing when the checksums match" do
+      expect(File).to receive(:read).with(md5_file_path).and_return(good_md5)
+      expect(subject).not_to receive(:cleanup_cached_tarball_path)
+      subject.verify_from_md5_file(md5_of_tarball)
+    end
+
+    it "raises an error and cleans up when the checksums do not match" do
+      expect(File).to receive(:read).with(md5_file_path).and_return(bad_md5)
+      expect(subject).to receive(:cleanup_cached_tarball_path)
+      expect(subject).to receive(:cleanup_md5_file_path)
+      expect { subject.verify_from_md5_file(md5_of_tarball) }.to raise_error(PuppetForge::V3::Release::ChecksumMismatch)
+    end
+  end
+
+  describe '#verify_from_forge' do
+    it "write the md5 to file when the checksums match" do
+      expect(subject.forge_release).to receive(:file_md5).and_return(good_md5)
+      expect(subject).not_to receive(:cleanup_cached_tarball_path)
+      expect(File).to receive(:write).with(md5_file_path, good_md5)
+      subject.verify_from_forge(md5_of_tarball)
+    end
+
+    it "raises an error and cleans up when the checksums do not match" do
+      expect(subject.forge_release).to receive(:file_md5).and_return(bad_md5)
+      expect(subject).to receive(:cleanup_cached_tarball_path)
+      expect { subject.verify_from_forge(md5_of_tarball) }.to raise_error(PuppetForge::V3::Release::ChecksumMismatch)
+    end
+    #it "verifies the module checksum based on the Forge file checksum" do
```

Comment:
Ah. Oops. That should be completely removed. That test is superseded by the "write the md5 to file when the checksums match" test. Removing.